File: W-21/Examples/heapsort.f
	subroutine HEAP_SORT(SEQUENCE, FIRST, LAST)
C
C Heap_Sort - Sort an array of items using a generalized heap-sort
C             technique.
	integer    SEQUENCE(*), FIRST, LAST
	integer    LEAF
	call BUILD_HEAP(SEQUENCE, FIRST, LAST)
	do LEAF = LAST, FIRST + 1, -1.
	   call SWAP(SEQUENCE(FIRST), SEQUENCE(LEAF))
	   call FIX_HEAP(SEQUENCE, FIRST, LEAF - 1)
	end do
	end

	subroutine BUILD_HEAP(SEQUENCE, FIRST, LAST)
C
C Build_Heap - Construct a heap from an array of items.
C
	integer   SEQUENCE(*), FIRST, LAST
	integer   FATHER
	do FATHER = (LAST / 2), FIRST, -1)
	   call FIX_HEAP(SEQUENCE, FATHER, LAST)
	end do
	end

	subroutine FIX_HEAP(SEQUENCE, FIRST, LAST)
C
C Fix_Heap - Repair a subheap.
C	integer    SEQUENCE(*), FIRST, LAST
	integer    FATHER, GREAT
	integer    MAX_CHILD
	logical    FOUND
	FOUND  = .false.
	FATHER = FIRST
	GREAT  = MAX_CHILD(SEQUENCE, FATHER, LAST)
	do while ((FATHER .le. (LAST / 2)) .and. .not. FOUND)
	   if (SEQUENCE(FATHER) .lt. SEQUENCE(GREAT)) then
	      call SWAP(SEQUENCE(FATHER), SEQUENCE(GREAT))
	      FATHER = GREAT
	      GREAT  = MAX_CHILD(SEQUENCE, FATHER, LAST)
	   else
	      FOUND = .true.
	   end if
	end do
	end

	integer function MAX_CHILD(SEQUENCE, FIRST, LAST)
C
C Max_Child - Find the largest of the two children of a node rooted at
C             FIRST.
C
	integer	SEQUENCE(*), FIRST, LAST
	integer LEFT, RIGHT
	LEFT  = FIRST * 2
	RIGHT = LEFT  + 1
	MAX_CHILD = LEFT
	if (RIGHT .le. LAST) then
	   if (SEQUENCE(RIGHT) .gt. SEQUENCE(LEFT)) then
	      MAX_CHILD = RIGHT
	   end if
	end if
	end

	subroutine SWAP(ARG_A, ARG_B)
C
C Swap - Swap two items.
C
	integer    ARG_A, ARG_B
	integer    TEMP
	TEMP  = ARG_A
	ARG_A = ARG_B
	ARG_B = TEMP
	end

File: W-21/Assignments/assignment5/pseudocode/heapsort.py
import random, sys

def max_child(A, first, last):
    left = first * 2
    right = left + 1
    child = left

    if right <= last and A[right] > A[left]:
        child = right

    return child

def fix_heap(A, first, last):
    found = False
    father = first
    great = max_child(A, father, last)

    while father <= last // 2 and not found:
        if A[father] < A[great]:
            A[father], A[great] = A[great], A[father]
            father = great
            great = max_child(A, father, last)
        else:
            found = True

def build_heap(A, first, last):
    for father in range(last // 2, first - 1, -1):
        fix_heap(A, father, last)

def heap_sort(A):
    first, last = 0, len(A) - 1
    build_heap(A, first, last)
    for leaf in range(last, first, -1):
        A[first], A[leaf] = A[leaf], A[first]
        fix_heap(A, first, leaf - 1)

for n in range(2, 1001):
    print(f"Sorting {n} elements... ", end='')
    arr = [random.randint(0, sys.maxsize) for _ in range(n)]
    heap_sort(arr)

    for i in range(1, len(arr)):
        if arr[i] < arr[i - 1]:
            print("failed")
            break
    else:
        print("passed")
